### PR TITLE
Invert function overloading order for AES-GCM doFinal decryption

### PR DESCRIPTION
**Pull request by geedo0**
*Issue #, if available:*
#271

*Description of changes:*
There are 3 overloaded methods for calling into a Cipher module which implement different ways of returning the output of the operation. From a memory management perspective these can be differentiated by whether or not the caller is responsible for allocating memory for the output.

Previously in our implementation, the top layer of the call stack is a method which allocates a buffer for the output and returns the result. For other interfaces in which the caller is responsible for allocating the output buffer we would allocate yet another buffer and copy it over to the buffer the caller gave us. This results in O(n) additional memory allocations and I/O.

This change inverts the call stack so that the top layer returns the result into an output buffer passed by the caller. That way, we only allocate a new output buffer when one is not provided. This gives us another 33% improvement in memory allocation from the benchmarks provided in issue #271.

n.b. Github is kinda inflexible about breaking up PRs so we scooped up the [pre-requisite PR](https://github.com/corretto/amazon-corretto-crypto-provider/pull/297) in this one. Once that's merged, I'll update this PR to omit the commit.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/com/amazon/corretto/crypto/provider/AesGcmSpi.java [hunk 1]
```diff
@@ -675,31 +680,21 @@ private byte[] engineDecryptFinal(byte[] input, final int offset, final int leng
...
+      workingInputArray = decryptInputBuf.getDataBuffer();
+      workingInputLength = decryptInputBuf.size();
+      workingInputOffset = 0;
 
       if (workingInputLength < tagLength) {
         throw new AEADBadTagException("Input too short - need tag");
       }
 
-      final int expectedOutputLength = workingInputLength - tagLength;
-      final byte[] result = new byte[expectedOutputLength];
```
Comment (geedo0) [inline src/com/amazon/corretto/crypto/provider/AesGcmSpi.java:701]: This is where the memory gets saved. Previously, when the caller provides the buffer this intermediate buffer gets allocated no matter what.

File: src/com/amazon/corretto/crypto/provider/AesGcmSpi.java [hunk 2]
```diff
@@ -675,31 +680,21 @@ private byte[] engineDecryptFinal(byte[] input, final int offset, final int leng
...
       final byte[] workingInputArray;
       final int workingInputOffset;
       final int workingInputLength;
-      if (decryptInputBuf.size() > 0) {
-        // We've already processed data so need to merge this in
-        engineUpdate(input, offset, length); // Decrypt mode never generates output for updates
-        workingInputArray = decryptInputBuf.getDataBuffer();
-        workingInputLength = decryptInputBuf.size();
-        workingInputOffset = 0;
-      } else {
```
Comment (geedo0) [inline src/com/amazon/corretto/crypto/provider/AesGcmSpi.java:690]: This is a small step backwards in favor of making things more uniform. Previously, we were able to get away with using the input buffer directly when the client calls doFinal using the byte[] interfaces and without calling doUpdate prior. With this change, internally buffering the input on decryption operations is now compulsory. If we have telemetry to suggest that this particular call pattern is common enough, we can try to add some logic to accommodate it. FWIW, Sun JSSE's TLS implementation calls into our library via the ByteBuffer interface where this has no impact.

File: src/com/amazon/corretto/crypto/provider/AesGcmSpi.java [hunk 3]
```diff
@@ -773,42 +762,37 @@ private byte[] engineDecryptFinal(byte[] input, final int offset, final int leng
...
-        throw new AssertionError(e);
-      }
-
-      if (actualLength == buf.length) {
-        return buf;
-      } else {
-        // This branch should never happen but is technically allowed by the underlying APIs.
-        // So, we cover it just in case.
-        return Arrays.copyOf(buf, actualLength);
+    final byte[] buf = new byte[engineGetOutputSize(length)];
```
Comment (geedo0) [inline src/com/amazon/corretto/crypto/provider/AesGcmSpi.java:765]: Allocation of the output buffer now happens here only in the "ACCP allocates the output buffer" interface. We avoid allocating the output buffer in the other two "caller allocates" interfaces.

File: src/com/amazon/corretto/crypto/provider/AesGcmSpi.java [hunk 4]
```diff
@@ -971,17 +955,22 @@ protected int engineDoFinal(final ByteBuffer input, final ByteBuffer output)
     return output.position() - initialPosition;
   }
 
-  private void checkOutputBuffer(final int inputLength, final byte[] output, final int outputOffset)
```
Comment (geedo0) [inline src/com/amazon/corretto/crypto/provider/AesGcmSpi.java:974]: These changes are all covered in the other PR. I couldn't figure out if Github supports commit slicing for PR.

https://github.com/corretto/amazon-corretto-crypto-provider/pull/297

File: src/com/amazon/corretto/crypto/provider/AesGcmSpi.java [hunk 5]
```diff
@@ -756,15 +748,12 @@ private byte[] engineDecryptFinal(byte[] input, final int offset, final int leng
...
-        return result;
-      } else if (outLen < result.length) {
-        // While technically allowed by the APIs, this should never happen.
-        return Arrays.copyOf(result, outLen);
-      } else {
-        throw new RuntimeCryptoException("Output longer than expected.");
-      }
+      return outLen;
+    } catch (AEADBadTagException e) {
+      int maxFillSize = output.length - outputOffset;
```
Comment (WillChilds-Klein) [inline src/com/amazon/corretto/crypto/provider/AesGcmSpi.java:753]: minor nit: `final`

**Review by WillChilds-Klein — APPROVED** [review]

**Review by amirhosv — APPROVED** [review]
What's the advantage of overriding the methods that process ByteBuffer? Those methods are not abstract and JDK provides [an implementation for them] (https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/javax/crypto/CipherSpi.java#L744)

**Pull request merged**